refs #562551 playwright documentation (#9)
easysoftware/developer-portal-devs: docs/Frontend_tutorials/FE-Testing.md

---

## Playwright tests

We are using Playwright for testing our Frontend.
In this section we will try to answer few questions, for example:
*How to write Playwright tests? How we manage them and how to run them?*

### Getting started with Playwright
At the beginning please read some basics about Playwright:

1. [Installing Playwright](https://playwright.dev/docs/intro#installing-playwright)
2. [Assertions](https://playwright.dev/docs/test-assertions)
3. And most importantly - [Best practices](https://playwright.dev/docs/selectors#best-practices).

### Structure

On the platform, Playwright tests are located in `/playwright` folder. The main configuration of Playwright is in `/playwright.config.ts` file.

Playwright itself runs JS tests in its own syntax against some running server - this is specified in `/playwright.config.ts`.

1. ./playwright/fixtures

> - This folder contains JSON files representing entities of our system. Both existing once and the new ones.

Reviewer: ```suggestion
- This folder contains JSON files representing entities in our system. If a JSON contains an ID the entity exists in the testing database, otherwise they're mostly used for filling inputs.
```

If HTML is rendered by backend you need to pass `"data-cy": "some-value"`, for example:

```ruby
link_to(l(:label), url, class: 'customize-button', "data-cy": "button__customize_page")
# OR
link_to(l(:label), url, class: 'customize-button', data: { cy: "button__customize_page" })
```
<!-- theme: warning -->
>This is only valid type of `data-cy` = in every other case it is forbidden due to consistency!

### How to run tests locally

Lets see something about Playwright.

> You need to run application server aside and run against it Playwright.

Reviewer: ```suggestion
You need to run an application server aside and run Playwright against it.
```

How to do that:
Make sure that you have bundle install and assets precompiled
1. Run
```
bundle install
```

2. Define new database for playwright. In the file `./config/database.yml` 
example:
```
development:
  adapter: mysql2
  encoding: utf8mb4
  database: playwright_test
  pool: 5
  username: playwright
  password:
```

3. Run
```
rails db:create
```

4. For testing, we use database from [https://cypresssource-minor.easysoftware.com](https://cypresssource-minor.easysoftware.com) .
   **It is forbidden to edit it without permission from QA!** We just simply dump the database and import dump into new database.

5. Run
```
bundle exec rake easyproject:install
```

6. Run your server aside in the test environment:
```ruby
bundle exec rails s
```

> FORCE_HTTP allows puma accept HTTP request (insecure) in production.

7. Run Playwright
```ruby
yarn playwright test
```

## Connection to GitlabCI – internal usage

> > At GitlabCI there is a playwright stage for running examples "automatically".

Reviewer: ```suggestion
> In GitlabCI there is a playwright stage for running examples "automatically".
```

> To execute a playwright test in GitlabCI we use docker image. For each branch, new docker image is being created, which is saved in Gitlab registers: https://git.easy.cz/devel/devel/container_registry.

Reviewer: ```suggestion
To execute playwright tests in GitlabCI we use docker images. For each branch, new docker image is created, which is saved in Gitlab registry: https://git.easy.cz/devel/devel/container_registry.
```

The name of the image contains the name of the branch for which it was created.

> Each failed pipeline provides an option to download detailed back-trace zip file in 'Job artifacts' section. You can execute this zip file after downloading it with `yarn playwright show-trace` command. 

Reviewer: ```suggestion
For debugging, each failed pipeline provides an option to download detailed backtrace zip file in 'Job artifacts' section. You can execute this zip file after downloading it with `yarn playwright show-trace` command. 
```
